First working  version of netconf adapter.

sjames-iberis opened the pull request.

sjames-iberis committed:
First working  version of netconf adapter.

SeanCondon commented on cmd/server.go:
```diff
@@ -15,19 +15,139 @@
+	key = serverCmd.Flags().String("key", "certs/localhost.key", "path to client private key")
+	cert = serverCmd.Flags().String("cert", "certs/localhost.crt", "path to client certificate")
+
+	port = serverCmd.Flags().Int("port", 11161, "port to listen")
```
we generally use 5150 as the default port (yes after the Van Halen album, but which was named after a California law)

SeanCondon commented on cmd/server.go:
```diff
@@ -15,19 +15,139 @@
+	port = serverCmd.Flags().Int("port", 11161, "port to listen")
+	isInsecure = serverCmd.Flags().Bool("insecure", false, "whether to enable skip verification")
+
+	deviceIP = serverCmd.Flags().String("device-ip", "10.228.63.5:830", "device ip address:port for NETCONF")
```
probably should not have a default here

SeanCondon commented on cmd/server.go:
```diff
@@ -15,19 +15,139 @@
+
+// RunGnmiServer provides an indirection so that the logic can be tested independently of the cobra infrastructure
+func RunGnmiServer(command *cobra.Command, args []string) {
+	log.Info("Run GNMI Server... ")
```
maybe feedback the port number

SeanCondon commented on cmd/version.go:
```diff
@@ -39,7 +39,7 @@ var versionCmd = &cobra.Command{
 
 func printVersion(w io.Writer, version, commit, date string) {
-	fmt.Fprintf(w, "%v, commit %v, built at %v \n", version, commit, date)
+	_, _ = fmt.Fprintf(w, "%v, commit %v, built at %v \n", version, commit, date)
```
I like this - we should adopt something like this in our modules

SeanCondon commented on get.all.gnmi:
```diff
@@ -0,0 +1,3 @@
+path: <
```
I really like these - I will start using the same style for `onos-config`

SeanCondon commented on gnmi_cli/README.md:
`````diff
@@ -0,0 +1,12 @@
+
+````
+gnmi_cli  \
+-address  localhost:10999   \
`````
maybe address should be `gnmi-netconf-adapter:5150` - as that should be the service name in k8s

SeanCondon commented on pkg/adapter/modeldata/gostruct/gen.go:
```diff
@@ -0,0 +1,17 @@
+
+package gostruct
+
+//go:generate sh -c "go get -u github.com/openconfig/ygot; (cd $GOPATH/src/github.com/openconfig/ygot && go get -t -d ./...); go get -u github.com/openconfig/public; go get -u github.com/YangModels/yang; cd $GOPATH/src && go run github.com/openconfig/ygot/generator/generator.go -generate_fakeroot -output_file github.com/google/gnxi/gnmi/modeldata/gostruct/generated.go -package_name gostruct -excl ...
```
If you use `go generate` from the Makefile it will create the "generated.go" file and it will not have to be versioned below.
I know there is an issue with the version of ygot, but this should be dealt with through the `go.mod`.

Also I should point out that the Docker base image we use "onosproject/golang-build:stable" has the "generator" from ygot in it - I think this is the right place to adjust the "ygot" version

SeanCondon reviewed (commented):
I got partly through it - set.go is next - I will resume in the morning

SeanCondon commented on pkg/adapter/capabilities.go:
```diff
@@ -0,0 +1,36 @@
+// Copyright 2019-present Open Networking Foundation.
```
year can be 2020. Same elsewhere

sjames-iberis replied:
Done.

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+	model = NewModel(modeldata.ModelData, gostruct.SchemaTree["Device"])
+)
+
+func TestCapabilities(t *testing.T) {
```
this could be broken out in to a separate file `capabilities_test.go` - would help make the files more manageable in size

sjames-iberis replied:
Done,

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+	ncResult    string
+}
+
+func TestGet(t *testing.T) {
```
this could be broken out in to a separate file `get_test.go` - would help make the files more manageable in size

sjames-iberis replied:
Done.

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+	}
+}
+
+func TestSubscribe(t *testing.T) {
```
this could be broken out in to a separate file `subscribe_test.go` - would help make the files more manageable in size

sjames-iberis replied:
Done.

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+// Copyright 2019-present Open Networking Foundation.
```
this file might be called `test_set.go` - with other test files corresponding to `subscribe.go`, `get.go` etc. It would help make the files more manageable in size

sjames-iberis replied:
Done

SeanCondon commented on pkg/adapter/modeldata/modeldata.go:
```diff
@@ -0,0 +1,33 @@
+//	openconfig-openflow 0.1.0,
+//	openconfig-platform 0.5.0,
+//	openconfig-system 0.2.0.
+package modeldata
```
I think the whole model plugin here could be loaded in to `gnmi_netconf_adapter` like the model plugins are now loaded in to `onos-config` i.e. as a sidecar container in Kubernetes. I think this is a good way that the adapter could reuse and be kept in sync with the model plugins that `onos-config` uses. The sidecar loads and copies its `'so'` file over to `onos-config` which loads it as a shared library.

SeanCondon replied:
For the Unit tests you could test with `TestDevice-1.0.0` which have a very wide range of YANG constructs. You could depend on this like
```
	td1 "github.com/onosproject/config-models/modelplugin/testdevice-1.0.0/testdevice_1_0_0"
	td2 "github.com/onosproject/config-models/modelplugin/testdevice-2.0.0/testdevice_2_0_0"
```

sjames-iberis replied:
Regarding the test models, I would prefer to hold off on this for the time being.
I agree that the test yang and generated model  is a useful resource; it certainly covers more than the junos fragment that we are currently using.
However, I can see that it would expose some of the known limitations in the current adapter implementation, relating to data type transformations; these are marked by TODOs in the code.
The TODOs will obviously need to be addressed in due course, but:
	•	For the time being I don’t think they compromise the value of the initial test/demo
	•	When they are implemented, I’d prefer to have access to a real schema/device combination to inform and validate the implementation

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/openconfig/gnmi/proto/gnmi"
```
i find `pb` a particularly meaningless and unintuitive alias for `gnmi` - I'd rather just see 'gnmi` used.

sjames-iberis replied:
Done, after renaming adapter package from gnmi to adapter.

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+		})
+
+	s, err := NewAdapter(model, mockNc)
+	if err != nil {
```
could use `assert.NoError(t, err, "error in creating server: %v", err)` here instead

sjames-iberis replied:
Done

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+
+	// Send request
+	resp, err := s.Get(context.TODO(), req)
+
```
same here

sjames-iberis replied:
Done

SeanCondon commented on pkg/adapter/adapter_test.go:
```diff
@@ -0,0 +1,688 @@
+	if !ok {
+		t.Fatal("got a non-grpc error from grpc call")
+	}
+	if gotRetStatus.Code() != tc.wantRetCode {
```
Could use `assert.Equal(t, tc.wantRetCode, gotRetStatus.Code(), "Unexpected return code")` here instead. The values will get printed in the error.

sjames-iberis replied:
Done

SeanCondon reviewed (approved):
This is really great work.

My comments are just suggestions - I'm OK to merge this as is, and if comments are to be addressed it can be done later (depending on the disposition)

sjames-iberis committed:
Address review comments:
- break up monolithic adapter test into operation-specific sources.
- update date in copyright notices.

sjames-iberis committed:
Address review comments:
- rename package from gnmi to adapter, then...
- ... pb import alias with gnmi

sjames-iberis committed:
Use assert to confirm adapter instantiation succeeded.

sjames-iberis committed:
Address review comments:
- use assertions to confirm error expectations.

SeanCondon commented:
@sjames-iberis I don't know if you saw the first set of comments I had left above e.g. using the port number 5150 etc. Anyway I want to merge this now and move on - if you're going to address these comments it can be done in a future patch

sjames-iberis commented:
@SeanCondon - I had been focused on the adapter package.  
I had seen and was not ignoring the comments on the cmd package and was intending to discuss with Kieran. Will follow up next week.  
Thanks for the feedback.
